### PR TITLE
Enable Parameterization via Notebook URI

### PR DESCRIPTION
**Pull request by VasuBhog**
This PR enables parameterization via Notebook URI as shown below.

The implementation covers how a user opens a new notebook via azuredatastudio://URI that contains parameters. 

ex. 
`URI with Parameter cell: azuredatastudio://microsoft.notebook/open?url=https://raw.githubusercontent.com/VasuBhog/AzureDataStudio-Notebooks/main/Demo_Parameterization/Input.ipynb?x=1&y=2`

Parameters will be parsed from the URI, then transformed to string format to then be added as a new cell to the notebook. The new cell will either be added as a _parameters tagged cell_ or as _injected parameters cell_ depending on if there is already a parameter cell in the notebook.

ParameterizationV2

### CODE REVIEW

File: extensions/notebook/src/protocol/notebookUriHandler.ts [hunk 1]
```diff
@@ -35,13 +34,13 @@ export class NotebookUriHandler implements vscode.UriHandler {
 	}
 
 	private open(uri: vscode.Uri): Promise<void> {
-		const data = querystring.parse(uri.query);
+		const data = uri.query.toString();
```
Comment (lucyzhang929) [inline extensions/notebook/src/protocol/notebookUriHandler.ts:37]: Just curious, why do we need these changes to the open method?

File: src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts [hunk 2]
```diff
@@ -318,6 +342,118 @@ suite('notebook model', function (): void {
...
+		assert.equal(model.cells.length, 2, 'Two cells should be entered');
+
+		// Set parameter cell and injected parameters cell
+		let notebookParamsCell = model.cells[0];
+		let notebookInjectedParamsCell = model.cells[1];
+
+		// Parameters Cell Validation
+		assert.equal(model.cells.indexOf(notebookParamsCell), 0, 'Notebook parameters cell should be first cell in notebook');
+		assert.equal(notebookParamsCell.isParameter, true, 'Notebook parameters cell should be tagged paramter');
+		assert.equal(notebookParamsCell.isInjectedParameter, false, 'Notebook parameters cell should be tagged paramter');
```
Comment (lucyzhang929) [inline src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts:365]: This description should say 'Notebook parameters cell should not be tagged injected parameter' right?

File: src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts [hunk 3]
```diff
@@ -318,6 +342,118 @@ suite('notebook model', function (): void {
...
+
+		// Injected Parameters Cell Validation
+		assert.equal(model.cells.indexOf(notebookInjectedParamsCell), 1, 'Notebook injected parameters cell should be first cell in notebook');
```
Comment (lucyzhang929) [inline src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts:368]: should be second cell

File: src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts [hunk 4]
```diff
@@ -318,6 +342,118 @@ suite('notebook model', function (): void {
...
+		assert.equal(model.notebookUri, defaultModelOptions.notebookUri, 'Notebook model has incorrect URI');
+		assert.equal(model.cells.length, 2, 'Two cells should be entered');
+
+		// Set parameter cell and injected parameters cell
+		let notebookParamsCell = model.cells[0];
+		let notebookInjectedParamsCell = model.cells[1];
+
+		// Parameters Cell Validation
+		assert.equal(model.cells.indexOf(notebookParamsCell), 0, 'Notebook parameters cell should be first cell in notebook');
+		assert.equal(notebookParamsCell.isParameter, true, 'Notebook parameters cell should be tagged paramter');
```
Comment (lucyzhang929) [inline src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts:364]: typo - parameters (I think it was copied to the other descriptions too)

File: src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts [hunk 5]
```diff
@@ -318,6 +342,118 @@ suite('notebook model', function (): void {
...
+		// When I initialize the model
+		let model = new NotebookModel(defaultModelOptions, undefined, logService, undefined, new NullAdsTelemetryService(), queryConnectionService.object, configurationService);
+		await model.loadContents();
+
+		assert.equal(model.notebookUri, defaultModelOptions.notebookUri, 'Notebook model has incorrect URI');
+		assert.equal(model.cells.length, 2, 'Two cells should be entered');
+
+		// Validate notebookUri parameter cell is set as injected parameter
+		let notebookUriParamsCell = model.cells[1];
+		assert.equal(model.cells.indexOf(notebookUriParamsCell), 1, 'NotebookURI parameters cell should be first cell in notebook');
```
Comment (lucyzhang929) [inline src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts:412]: should be second cell

File: src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts [hunk 6]
```diff
@@ -318,6 +342,118 @@ suite('notebook model', function (): void {
...
+		assert.equal(notebookUriParamsCell.isParameter, false, 'NotebookURI parameters cell should be tagged paramter');
```
Comment (lucyzhang929) [inline src/sql/workbench/contrib/notebook/test/electron-browser/notebookModel.test.ts:413]: Think these descriptions are flipped (should not be tagged parameter and should be injected parameter)

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 7]
```diff
@@ -363,9 +363,19 @@ export class NotebookModel extends Disposable implements INotebookModel {
 						}
 					});
 				}
+				// Get Notebook URI Params & adjust to string
+				let notebookUriParams: string = this.notebookUri?.query;
+				notebookUriParams = notebookUriParams.split('').join(' ').replace('& ', '\n');
+				// Get parameter cell and index to place new notebookUri parameters accordingly
+				let parameterCellIndex = 0;
+				let parameterCell = false;
```
Comment (lucyzhang929) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:371]: hasParameterCell might be a better variable name

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 1 above)
Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: So the one issue I found when passing in URIs with `?x=1&y=2` is that it will cause the data variable to be corrupted and lose the query part of the URI. This was a way where we keep URI as a string entirely and not as a querystring.

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts (hunk 7 above)
Comment (VasuBhog) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:371]: Yeah, that would make it clear this notebook _has_ a parameter cell.

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 1 above)
Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: @lucyzhang929. Here is a quick comparison. The use of querystring is not able to handle the parameters query format that gets added to the end of URI for parameterization. Instead, `oldData` (shown below) gets set as a dictionary object which would then fail the openNotebook decoding of the URL and not get the proper parameters.

image
image

Comment (Charles-Gagnon) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: Can you give an example of the full URI passed in to `handleUri`?

Because it should look something like this : `azuredatastudio://microsoft.notebook/open?url=https://raw.githubusercontent.com/microsoft/sqlworkshops/master/sql2019lab/01_IntelligentPerformance/iqp/iqp_tablevariabledeferred.ipynb&x=1&y=2`

Note that x and y are values in the query string along with the URL - they are NOT query parameters on the notebook URL.

If done like this then I would expect the current approach to work fine.

Although if you have an actual reason for wanting to embed the query paramters into the URL itself then what should also work is to parse that URL separately - something like this : 

`const notebookUri = vscode.Uri.parse(uri.query.toString());`

and then pass that in and use notebookUri.query to access the query parameters at that point.

File: extensions/notebook/src/protocol/notebookUriHandler.ts [hunk 8]
```diff
@@ -70,14 +69,15 @@ export class NotebookUriHandler implements vscode.UriHandler {
 			}
 
 			let contents = await this.download(url);
-			let untitledUri = this.getUntitledUri(path.basename(uri.fsPath));
+			let untitledUriPath = this.getUntitledUriPath(path.basename(uri.fsPath));
+			let untitledUri = uri.with({ authority: '', scheme: 'untitled', path: untitledUriPath });
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/protocol/notebookUriHandler.ts:73]: Why are you changing this? You don't seem to use `untitledUri` anywhere now

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 9]
```diff
@@ -363,9 +363,20 @@ export class NotebookModel extends Disposable implements INotebookModel {
 						}
 					});
 				}
+				// Get Notebook URI Params & adjust to string
+				let notebookUriParams: string = this.notebookUri?.query;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:367]: Why the ? - it looks like the URI is never undefined?

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 10]
```diff
@@ -450,6 +466,28 @@ export class NotebookModel extends Disposable implements INotebookModel {
...
 
+	// Adds Paramters cell based on Notebook URI parameters
+	private addParametersCell(notebookUriParams: string, hasParameterCell: boolean, parameterCellIndex: number): void {
+		let uriParamsIndex = parameterCellIndex;
+		// Set new parameters as Injected Parameters cell after original parameter cell
+		if (hasParameterCell) {
+			uriParamsIndex = parameterCellIndex + 1;
+			this.addUriParameterCell(uriParamsIndex, notebookUriParams);
+			this.cells[uriParamsIndex].isInjectedParameter = true;
+			this.cells[uriParamsIndex].source = ['# Injected-Parameters\n'].concat(this.cells[uriParamsIndex].source);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:477]: Localize?

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 11]
```diff
@@ -450,6 +466,28 @@ export class NotebookModel extends Disposable implements INotebookModel {
...
+		} else {
+			// Set new parameters as the parameters cell as the first cell in the notebook
+			this.addUriParameterCell(uriParamsIndex, notebookUriParams);
+			this.cells[uriParamsIndex].isParameter = true;
+		}
+	}
+
+	// Add new notebook URI parameters cell to notebook
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:485]: Use actual js doc (///) comments for all function comments please

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 1 above)
Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: Sure, so I do give an example of full URI actually in the output shown. This is shown in the gif as well.

`azuredatastudio://microsoft.notebook/open?https://raw.githubusercontent.com/VasuBhog/AzureDataStudio-Notebooks/main/Demo_Parameterization/Input.ipynb?x=1&y=2`

The reason to embed the query into the URI is so that when we load the new notebook in NotebookModel, we are able to utilize the URI to get the parameters notebookUri.query. I discussed this with @chlafreniere before the holidays and this was a solid route that enables both the proper setup of the URI for more generic use later in the future as well. 

We absolutely can not use the querystring method. The second part of what you said does not make sense. 
`const notebookUri = vscode.Uri.parse(uri.query.toString());` this would mean we setting the uri only to the query parameters. The above line is specifically ensuring we are getting the URI properly. If you look at the next function after `openNotebook` you will see we do validate this Uri where we actually set the Uri similar to how you have but in the next function.

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 8 above)
Comment (Charles-Gagnon) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:73]: Ah whoops, guess I didn't expand it or something.

I'm still not understanding why this is necessary though. What happens if you leave it as it was?

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts (hunk 9 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:367]: If it can be undefined then you should be updating the property getter to reflect that - right now it's just returning a URI. Make sure you keep an eye out for that kind of stuff here when making changes - we don't currently strictly enforce this in a lot of places but it doesn't mean we shouldn't keep the typings accurate when we have the chance.

But in this case I was specifically talking about the URI being undefined, NOT the query parameters (which I agree can be empty/undefined and should be handled as such). A notebook should always have a URI associated with it and indeed we appear to do a check for this in the constructor

```
if (!_notebookOptions || !_notebookOptions.notebookUri || !_notebookOptions.notebookManagers) {
	throw new Error('path or notebook service not defined');
}
```

so doing this should be fine here : 

`const notebookUriParams: string = this.notebookUri.query;`

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts (hunk 10 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:477]: The `Injected- Parameters` text

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts (hunk 11 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:485]: It is and should be - you don't get intellisense with just normal // comments so that kind of defeats the purpose of having function doc comments in the first place. 

(and yeah it's /**, not ///. Darn C# conventions messing me up again)

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 1 above)
Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: Ah I overlooked the URL part thanks for pointing that out. I will be sure to test out my work with this and adjust it to ensure this version works.

Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: > The query parameter string should be a set of key/value pairs. Your new code though is changing this so the top level query string is another full URL itself. This is not only meaning that previous links which used the old url= format are breaking but is also not following the standard URI format.

@Charles-Gagnon I believe that the data variable (set to uri.query.toString) can not be key/value pairs for the specific reason I showed in my quick comparison above (oldData vs data). If you see the way oldData is established it uses the key and value pairs of the URI, if I were to change it back to querystring and have the appropriate standard URI format (uri.query.toString for data) then the **parameters** that is injected will be misformatted and then we will have another error regarding opening the URL and it will throw the error `Failed to open URI`.

Moreover, `const notebookUri = vscode.Uri.parse(uri.query.toString());` gives the Error: _[UriError]: Scheme contains illegal characters._ since uri.query is `url=https://raw.githubusercontent.com/VasuBhog/AzureDataStudio-Notebooks/main/Demo_Parameterization/Input.ipynb`.

I think I would be able to set data like this - 
`const data = uri.query.toString();` which would then store "url=https://raw.githubusercontent.com/VasuBhog/AzureDataStudio-Notebooks/main/Demo_Parameterization/Input.ipynb" and then I can parse for everything after the url=. What are your thoughts on this?

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 8 above)
Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:73]: If I don't do anything then it won't open the new notebook since the authority needs to be undefined for the URI.

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 1 above)
Comment (Charles-Gagnon) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: So yeah - my example wouldn't work directly since I was basing it off of your changes. 

Here's what I'm envisioning this will look like : 

1. You keep the parsing logic as it is 

`const data = querystring.parse(uri.query);`

This will end up with an object that has a single url property containing the full URL (including parameters) - it should be something like this : 

`{ url : 'https://raw.githubusercontent.com/VasuBhog/AzureDataStudio-Notebooks/main/Demo_Parameterization/Input.ipynb?x=1&y=2' }`

Which you'll then pass to openNotebook as was being done before.

But as I mentioned earlier I'm pretty sure you're going to have to encode the parameters here since you don't want them messing with the top level URI. So at some point you're likely going to have to decode that as well, which you probably want to do here. Something like `decodeURIComponent` should do the trick.

Comment (Charles-Gagnon) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: It actually looks like we already decode the URL on line 51 so you probably don't even need to do anything else for that.

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 8 above)
Comment (Charles-Gagnon) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:73]: That error isn't really telling me what's going on - you're saying you got that error when passing in the uri object like the code was doing before? Why would that be? The only thing that should be changing would be the parameters being added so I'm not sure why that would break this...

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 1 above)
Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: > `const data = querystring.parse(uri.query);`

So I have the exact example of how this will look with parameters query (discussed after Lucy's comment) added to URI ie '...Input.ipynb?x=1&y=2'. As I said we can not use `querystring.parse` shown below:

image
image

As you can see the querystring version would make key and value pairs for the **parameters**. I think having my version of this data is generic and works for the standard URL format. 

This will then be decoded properly and we are able to ensure the parameters query for this URI (?x=1&y=2) is set to the URI.

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 8 above)
Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:73]: I think the best way to answer most of your questions would be to show you offline. It will be clearly visible through the debugging experience.

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 1 above)
Comment (Charles-Gagnon) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: That's why you need to have the caller encode the URL - to avoid the parameters being seen as part of the top level URI (the azuredatastudio:// part with the url parameter). So the full URL would be something like this : 

`azuredatastudio://microsoft.notebook/open?url=https%3A%2F%2Fraw.githubusercontent.com%2FVasuBhog%2FAzureDataStudio-Notebooks%2Fmain%2FDemo_Parameterization%2FInput.ipynb%3Fx%3D1%26y%3D2`

You then pull out the url value : `https%3A%2F%2Fraw.githubusercontent.com%2FVasuBhog%2FAzureDataStudio-Notebooks%2Fmain%2FDemo_Parameterization%2FInput.ipynb%3Fx%3D1%26y%3D2` and pass it through decodeURIComponent to get the actual URL to open : `https://raw.githubusercontent.com/VasuBhog/AzureDataStudio-Notebooks/main/Demo_Parameterization/Input.ipynb?x=1&y=2` which can then be parsed as a normal URI to get the query parameters.

Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: After discussions with @Charles-Gagnon, it seems that the encoded URI with parameters is already decoded from the extHostUrls (VSCode side). As a result they are decoded by the time we get the uri.query back and set as a string. 

For now I will push my update using the split method, but we will discuss further with @chlafreniere.

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 8 above)
Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:73]: Discussed this as well with @Charles-Gagnon, for now, this can be handled here, but a suggestion would be to move it to the handleUri function.

Comment (Charles-Gagnon) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:73]: @VasuBhog I wasn't talking about this logic - this stuff is fine to keep here. It's the other code to parse out the notebook URL + parameters that I was saying I'd do in the handleUri function.

File: extensions/notebook/src/protocol/notebookUriHandler.ts [hunk 12]
```diff
@@ -126,6 +126,6 @@ export class NotebookUriHandler implements vscode.UriHandler {
 			}
 			nextVal++;
 		}
-		return vscode.Uri.parse(`untitled:${title}`);
+		return `${title}`;
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/protocol/notebookUriHandler.ts:129]: Just return title - no reason to wrap this in an interpolated string

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 13]
```diff
@@ -363,23 +363,41 @@ export class NotebookModel extends Disposable implements INotebookModel {
...
 				}
+				// Get Notebook URI Params & adjust to string
+				let notebookUriParams: string = this.notebookUri.query;
+				// Format query to code cell format
+				notebookUriParams = notebookUriParams.replace(/&/g, '\n').replace(/=/g, ' = ');
+				// Get parameter cell and index to place new notebookUri parameters accordingly
+				let parameterCellIndex = 0;
+				let hasParameterCell = false;
+				let hasInjectedCell = false;
+				let injectedParametersMsg = localize('injectedParametersMsg', '# Injected-Parameters\n');
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:374]: Put this as a const in the file so you don't have to evaluate it every time this function is called

Comment (Charles-Gagnon) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:374]: And that way you can also re-use it instead of redefining it below

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 14]
```diff
@@ -450,6 +469,33 @@ export class NotebookModel extends Disposable implements INotebookModel {
 		return cell;
 	}
 
+	/* Adds Paramters cell based on Notebook URI parameters */
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:472]: This should have the parameters too per normal js doc formatting - if you type `/**` and hit enter in the line above the function it should generate the template for you

Comment (VasuBhog) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:472]: gotcha thanks for letting me know.

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts (hunk 13 above)
Comment (VasuBhog) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:374]: Will do. Thanks!

File: extensions/notebook/src/protocol/notebookUriHandler.ts [hunk 15]
```diff
@@ -35,13 +34,13 @@ export class NotebookUriHandler implements vscode.UriHandler {
 	}
 
 	private open(uri: vscode.Uri): Promise<void> {
-		const data = querystring.parse(uri.query);
+		const data = uri.query.toString().split('url=')[1];
```
Comment (corivera) [inline extensions/notebook/src/protocol/notebookUriHandler.ts:37]: Wouldn't this fail if the query string doesn't have a url parameter? You'd end up with a length 1 array after the split.

Comment (Charles-Gagnon) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: We should always have a url parameter so this shouldn't be a problem for valid URIs.

But you do need to account for url= being in folder/file names. So if you're going this route you should just find the first instance of that and split on that - not split the whole string.

Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: The azuredatastudio URI should follow standard URI standard. As a result, any link that a user clicks on should always have the 'url='.
`azuredatastudio://microsoft.notebook/open?url=`

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 16]
```diff
@@ -363,23 +364,40 @@ export class NotebookModel extends Disposable implements INotebookModel {
 						}
 					});
 				}
+				// Get Notebook URI Params & adjust to string
+				let notebookUriParams: string = this.notebookUri.query;
+				// Format query to code cell format
+				notebookUriParams = notebookUriParams.replace(/&/g, '\n').replace(/=/g, ' = ');
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:370]: You should test what happens when the parameters have & or = in them (both key and value). Ideally you should be able to have those values be % encoded before hand and then decode them here - but given what we found here we may not be able to do this without core changes. But at the very least you should test this out and see what the behavior is so we can at least document it.

Comment (VasuBhog) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:370]: Interesting, are you saying something like `...Input.ipynb?x=1&1&y=2`. 
I believe that this is not a valid scenario for parameterization. Moreover, would be an extreme edge case in which we would not expect users to ever use **&** within parameters.

Comment (Charles-Gagnon) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:370]: Why is that not a valid scenario? I would expect string parameters to be supported as well. And I wouldn't necessarily call an & in a string an "extreme edge case".

But that's up to you to determine if you want to support. Just make sure if you don't it's clearly documented. And I would suggest bringing this up with the other stakeholders in this spec so they're aware what is/isn't supported.

Comment (Charles-Gagnon) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:370]: And generally I would suggest that if you do have edge cases like this that you handle in certain ways you add tests for those cases. This shows

1. That the behavior is as you expect when you wrote the code originally
2. People can clearly see what is supported and how it should behave in case they're making changes later

Comment (VasuBhog) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:370]: Yup, I think that would be a difficult edge case to solve as well, `...Input.ipynb?x=1&1&y=2`. Since we specifically use this format with **&** as a separator for Notebook Parameters we would treat them as a new parameter.
As such, I will be sure to document this. I can create more test cases to ensure that this is the behavior we expect.

File: extensions/notebook/src/protocol/notebookUriHandler.ts [hunk 17]
```diff
@@ -35,13 +34,13 @@ export class NotebookUriHandler implements vscode.UriHandler {
 	}
 
 	private open(uri: vscode.Uri): Promise<void> {
-		const data = querystring.parse(uri.query);
+		const data = uri.query.substr(uri.query.indexOf('url=') + 4);
```
[the remark this answers is not in the document]
Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts]: Let me know if I should add more descriptions :)

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 18]
```diff
@@ -450,6 +469,32 @@ export class NotebookModel extends Disposable implements INotebookModel {
...
+			this.cells[uriParamsIndex].source = [injectedParametersMsg].concat(this.cells[uriParamsIndex].source);
+		} else {
+			// Set new parameters as the parameters cell as the first cell in the notebook
+			this.addUriParameterCell(uriParamsIndex, notebookUriParams);
+			this.cells[uriParamsIndex].isParameter = true;
+		}
+	}
+
+	/** Add new notebook URI parameters cell to notebook */
+	private addUriParameterCell(uriParamsIndex, notebookUriParams): void {
```
Comment (lucyzhang929) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:493]: Missing types for method parameters.

Comment (lucyzhang929) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:493]: Having both addParametersCell and addUriParametersCell methods seems a bit confusing imo since their method descriptions are the same. Maybe it'd be better to remove this helper method and rename the top one to addUriParameterCell since that is more descriptive? Then injectedParametersMsg and notebookUriParams can be set as the source at the same time (line 484).

**Review by lucyzhang929 — APPROVED** [review]

File: extensions/notebook/src/protocol/notebookUriHandler.ts [hunk 19]
```diff
@@ -33,15 +32,25 @@ export class NotebookUriHandler implements vscode.UriHandler {
 				vscode.window.showErrorMessage(localize('notebook.unsupportedAction', "Action {0} is not supported for this handler", uri.path));
 		}
 	}
-
+	/**
+	 * Our Azuredatastudio URIs follow the standard URI format of
+	 * azuredatastudio://microsoft.notebook/open?url=...
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/protocol/notebookUriHandler.ts:37]: This should really : 

1. Give examples of actual full URIs
2. Mention that it supports parameters as well and show examples of that
3. Mention that it only supports http(s) links

File: extensions/notebook/src/protocol/notebookUriHandler.ts [hunk 20]
```diff
@@ -33,15 +32,25 @@ export class NotebookUriHandler implements vscode.UriHandler {
...
+		// We ensure that the URI is formatted properly
+		let urlIndex = uri.query.indexOf('url=');
+		if (urlIndex > 0) {
+			// Querystring can not be used as it incorrectly turns parameters attached
+			// to the URI query into key/value pairs and would then fail to open the URI
+			data = uri.query.substr(urlIndex + 4);
+		}
 
-		if (!data.url) {
+		if (!data) {
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/protocol/notebookUriHandler.ts:49]: Just noticed this isn't returning when this happens - which it should be since this is an error case.

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 21]
```diff
@@ -450,6 +469,32 @@ export class NotebookModel extends Disposable implements INotebookModel {
 		return cell;
 	}
 
+	/** Adds Paramters cell based on Notebook URI parameters */
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:472]: nit: The typical style puts the comment on multiple lines : 

```
/**
* Adds Paramters cell based on Notebook URI parameters
*/
```

(you also misspelled parameters. If you don't already I would suggest installing a spell check extension in vs code to help avoid those)

Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:472]: These two functions still don't have js doc parameters

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 22]
```diff
@@ -450,6 +469,32 @@ export class NotebookModel extends Disposable implements INotebookModel {
...
+			if (hasInjectedCell) {
+				uriParamsIndex = uriParamsIndex + 1;
+			}
+			this.addUriParameterCell(uriParamsIndex, notebookUriParams);
+			this.cells[uriParamsIndex].isInjectedParameter = true;
+			this.cells[uriParamsIndex].source = [injectedParametersMsg].concat(this.cells[uriParamsIndex].source);
+		} else {
+			// Set new parameters as the parameters cell as the first cell in the notebook
+			this.addUriParameterCell(uriParamsIndex, notebookUriParams);
+			this.cells[uriParamsIndex].isParameter = true;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:488]: So if there's already a parameters cell then we create a new cell that's an "injected parameters" cell. But if we don't then we create a normal parameters cell? Why the difference? I would expect that we either always add the injected parameters into a parameters cell or we create a new injected parameters cell each time.

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 23]
```diff
@@ -363,23 +364,40 @@ export class NotebookModel extends Disposable implements INotebookModel {
 						}
 					});
 				}
+				// Get Notebook URI Params & adjust to string
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:367]: Not sure what the "adjust to string" means

File: extensions/notebook/src/protocol/notebookUriHandler.ts [hunk 24]
```diff
@@ -33,15 +32,25 @@ export class NotebookUriHandler implements vscode.UriHandler {
...
 		}
 	}
-
+	/**
+	 * Our Azuredatastudio URIs follow the standard URI format of
+	 * azuredatastudio://microsoft.notebook/open?url=...
+	 */
 	private open(uri: vscode.Uri): Promise<void> {
-		const data = querystring.parse(uri.query);
+		let data: string;
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/protocol/notebookUriHandler.ts:40]: Would love to see some more tests added covering the new scenarios you're supporting

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 19 above)
Comment (Charles-Gagnon) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: When writing comments always try to approach it from the point of view of someone who isn't intimately familiar with the code - and so outlining specific formats expected/supported is key to making it easy for other people to quickly and easily understand what is supposed to be having. (with tests backing it up to ensure that functionality is being kept the same)

Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:37]: Totally agree. It is quite interesting to me that there were about zero comments from when I started this, was this never recommended earlier on? I think it is a fantastic idea to give thorough comments, as this will help engineers of all types.

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts (hunk 22 above)
Comment (VasuBhog) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:488]: This is currently the designed plan from what has been discussed. 
The reason for this is that if a user goes from a normal notebook with absolutely zero parameters, and would like to turn this notebook (hosted somewhere) into a parameterized notebook (properly) then we should create a normal parameters cell (with the visual tag). As a result, the user can use this now as their "original" parameterized notebook. They can then host this, and continue chaining of parameterized notebooks (this work will enable this future work) with always having the original parameters notebook normal parameter cell noted.

File: extensions/notebook/src/protocol/notebookUriHandler.ts (hunk 24 above)
Comment (VasuBhog) [reply extensions/notebook/src/protocol/notebookUriHandler.ts:40]: Yes, I can do that. I would like to work on tests on Testing Monday moreover for these specific scenarios and begin working on getting this PR in so that I can work with Stuart on his Executable TSGs.

**Review by Charles-Gagnon — APPROVED** [review]
Lots of general comments I would love to see addressed but overall I don't see any specific issues with the implementation. I do still think that further investigation into properly encoding/decoding the URI to avoid the custom parsing (url=) would be great but it's not really an functionality blocker or anything.

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts (hunk 23 above)
Comment (VasuBhog) [reply src/sql/workbench/services/notebook/browser/models/notebookModel.ts:367]: // Modify Notebook URI Params format from URI query to string space delimited format
